Guide for developers (#165)
flanksource/canary-checker: docs/dev-guide.md

# Dev Guide

> The guide provides steps to configure your local setup with running canary-checker

Reviewer: Please add full stops to the end of lines.

> ## Setup a Kubernetes Cluster

Reviewer: ```suggestion
## Set up a Kubernetes Cluster
```

"Setup" is a noun. "Set up" is the verb form.

Before proceeding one must have access to a functioning Kubernetes cluster

In this guide we'll be working with [kind](https://kind.sigs.k8s.io/)

### Create a cluster

- Install latest version of [kind](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
- Create the cluster with `kind create cluster --name <cluster-name>`

### Install Metrics Server and Prometheus

For installing and configuring metrics server and prometheus please see the [prereqs](prereqs.md)

> 

Reviewer: Please keep the spaces between paragraphs constant. At some stage I'll move us to a linting tool called markdownlint and we'll get automation on stuff like this.

## Install Node

> The built-in dashboard of canary-checker uses node v12. So developer need to install the compatible version

Reviewer 2: Developers

> The built-in dashboard of canary-checker uses node v12. So developers need to install the compatible version

Reviewer: ```suggestion
The built-in dashboard of canary-checker uses node v12. Developers will need to install the compatible version:
```

> - Install [nvm](https://github.com/nvm-sh/nvm) to install and manage different version of node

Reviewer 2: Versions

- After nvm is installed one can install node v12 by running `nvm install v12`

## Build the canary-checker binary

> In most of the development work it makes sense to just build the binary and run th operator directly. Compared to the alternative approach of building the container image and using it

Reviewer 2: the operator

> In most of the development work it makes sense to just build the binary and run the operator directly. Compared to the alternative approach of building the container image and using it

Reviewer: Space between heading and text, please.

> In most of the development work it makes sense to just build the binary and run the operator directly. Compared to the alternative approach of building the container image and using it

Reviewer: ```suggestion
For most development work, it makes sense to build the binary and run the operator directly, rather than building and running a container image. With this in mind, this is how we can build and use the binary:
```

- Based on your current distro build the binaries: `make linux|darwin-arm64|darwin-amd64|windows`

> - After building the binary one needs to apply the canary-checker CRD inside the cluster. `kubectl apply -f config/deploy/crd.yaml`

Reviewer: ```suggestion
- After building the binary one needs to apply the canary-checker CRD inside the cluster: `kubectl apply -f config/deploy/crd.yaml`
```

- Once the CRDs are present in the cluster one can start the operator using the binary by: `./.bin/canary-checker-amd64 operator`
The above command will also start the canary-dashboard on `0.0.0.0:8080`

> Now you can deploy the canary-checks in any namespace of the cluster and they'll be reconciled by the operator  

Reviewer: I'd give examples here for useful checks to deploy and how to tell that they have been deployed.

> Note: The above command needs to be run from the root of canary-checker repo

Reviewer: ```suggestion
**Note:** The above command needs to be run from the root of canary-checker repo
```

Might be worth rephrasing this and putting it above the command.

# Dev Guide

The guide provides steps to configure your local setup with running canary-checker.

## Set up a Kubernetes Cluster

Before proceeding one must have access to a functioning Kubernetes cluster.

In this guide we'll be working with [kind](https://kind.sigs.k8s.io/).

### Create a cluster

- Install latest version of [kind](https://kind.sigs.k8s.io/docs/user/quick-start/#installation).
- Create the cluster with `kind create cluster --name <cluster-name>`.

### Install Metrics Server and Prometheus

For installing and configuring metrics server and prometheus please see the [prereqs](prereqs.md).

## Install Node

The built-in dashboard of canary-checker uses node v12. Developers will need to install the compatible version:
- Install [nvm](https://github.com/nvm-sh/nvm) to install and manage different versions of node.
- After nvm is installed one can install node v12 by running `nvm install v12`.

## Build the canary-checker binary

> **Note**: The following commands needs to be run from the root of canary-checker repo

Reviewer: Needs fullstop.

> **Note**: The following commands needs to be run from the root of canary-checker repo

Author: ```suggestion
**Note**: The following commands needs to be run from the root of canary-checker repo.
```

For most development work, it makes sense to build the binary and run the operator directly, rather than building and running a container image. With this in mind, this is how we can build and use the binary:
- Based on your current distro build the binaries: `make linux|darwin-arm64|darwin-amd64|windows`
- After building the binary one needs to apply the canary-checker CRD inside the cluster: `kubectl apply -f config/deploy/crd.yaml`
- Once the CRDs are present in the cluster one can start the operator using the binary by: `./.bin/canary-checker-amd64 operator`

The above command will also start the canary-dashboard on `0.0.0.0:8080`.

> Now you can deploy the canary-checks in any namespace of the cluster and they'll be reconciled by the operator  .

Reviewer: Too many spaces at end of line before .

> Now you can deploy the canary-checks in any namespace of the cluster and they'll be reconciled by the operator  .

Author: ```suggestion
Now you can deploy the canary-checks in any namespace of the cluster and they'll be reconciled by the operator.
```

> You can test if your operator is working correctly by deploying a sample Canary by running the command given below:

Reviewer: ```suggestion
You can test if your operator is working correctly by deploying a sample Canary (by `kubectl` applying the following template):
```

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: canaries.flanksource.com/v1
kind: Canary
metadata:
  name: http-pass
spec:
  interval: 30
  http:
    - endpoint: https://httpstat.us/200
      thresholdMillis: 3000
      responseCodes: [201, 200, 301]
      responseContent: ""
      maxSSLExpiry: 7
EOF
```

You can test the canary status by running: `kubectl get canaries.canaries.flanksource.com`

Sample output:
```
kubectl get canaries.canaries.flanksource.com
NAME        INTERVAL   STATUS   MESSAGE   UPTIME 1H    LATENCY 1H   LAST TRANSITIONED   LAST CHECK
http-pass   30         Passed             1/1 (100%)   500ms                            7s
```

> The above canary will also start showing on the dashboard(0.0.0.0:8080).:

Reviewer: ```suggestion
The above canary will also start showing on the dashboard (at `0.0.0.0:8080`):
```
